polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -29,297 +32,548 @@ namespace {
-              conflict = true;
-              return WalkResult::interrupt();
-            }
-            return WalkResult::advance();
-          });
-        }
+//===----------------------------------------------------------------------===//
+// LoopGuardBuilder
+//===----------------------------------------------------------------------===//
 
-        return conflict;
-      };
+std::unique_ptr<LoopGuardBuilder>
+LoopGuardBuilder::create(LoopLikeOpInterface loop) {
+  return TypeSwitch<Operation *, std::unique_ptr<LoopGuardBuilder>>(
+             (Operation *)loop)
+      .Case<scf::ForOp>(
+          [](auto loop) { return std::make_unique<SCFForGuardBuilder>(loop); })
+      .Case<scf::ParallelOp>([](auto loop) {
+        return std::make_unique<SCFParallelGuardBuilder>(loop);
+      })
+      .Case<AffineForOp>([](auto loop) {
+        return std::make_unique<AffineForGuardBuilder>(loop);
+      })
+      .Case<AffineParallelOp>([](auto loop) {
+        return std::make_unique<AffineParallelGuardBuilder>(loop);
+      })
+      .Default([](auto) { return nullptr; });
+}
 
-  // If the operation has side effects, check whether other operation in the
-  // loop prevent hosting it.
-  if ((memoryEffects.readsFromMemory() || memoryEffects.writesToMemory() ||
-       memoryEffects.freesMemory()) &&
-      hasConflictsInLoop(op, loop)) {
-    LLVM_DEBUG(llvm::dbgs().indent(2)
-               << "cannot be hoisted: found conflicting operation\n");
-    return false;
-  }
+//===----------------------------------------------------------------------===//
+// SCFForLoopGuardBuilder
+//===----------------------------------------------------------------------===//
 
-  // Recurse into the regions for this op and check whether the contained ops
-  // can be hoisted. We can inductively assume that this op will have its block
-  // args available outside the loop.
-  SmallPtrSet<Operation *, 2> willBeMoved2(willBeMoved.begin(),
-                                           willBeMoved.end());
-  willBeMoved2.insert(&op);
+void SCFForGuardBuilder::guardLoop() const {
+  scf::IfOp ifOp = createGuard();
+  loop->moveBefore(ifOp.thenYield());
+  replaceUsesOfLoopReturnValues(ifOp);
 
-  for (Region &region : op.getRegions()) {
-    for (Operation &innerOp : region.getOps()) {
-      if (!canBeHoisted(innerOp, loop, willBeMoved2))
-        return false;
-      willBeMoved2.insert(&innerOp);
-    }
-  }
+  bool yieldsResults = !loop->getResults().empty();
+  if (!yieldsResults)
+    ifOp.elseBlock()->erase();
+}
 
-  LLVM_DEBUG(llvm::dbgs().indent(2) << "can be hoisted: no conflicts found\n");
+Value SCFForGuardBuilder::createGuardExpr() const {
+  return builder.create<arith::CmpIOp>(
+      loop.getLoc(), arith::CmpIPredicate::sle,
+      builder.create<arith::AddIOp>(loop.getLoc(), loop.getLowerBound(),
+                                    loop.getStep()),
+      loop.getUpperBound());
```

Comment:
why is it not 
```suggestion
      loop.getLoc(), arith::CmpIPredicate::slt,
      loop.getLoc(), loop.getLowerBound(),
      loop.getUpperBound());
```
same question to the other kind of loop.

Reply:
done